Upgrade Ruma

jplatte opened the pull request:
I finished the dynamically-sized identifiers work. It was a lot. Now the upgrade turns out to also be a lot. I think some of the symbol soup could be gotten rid of again if `Arc` was used more consistently. Currently lots of `Arc<FooId>` <=> `Box<FooId>` conversions are being done.

jplatte commented:
Hope it's okay to ignore `clippy::type_complexity` in all the new places it's popped up.

jplatte commented:
CI fails with

```
error: implementation of `std::ops::FnOnce` is not general enough
   --> crates/matrix-sdk-appservice/src/webserver/warp.rs:102:14
    |
102 |             .and_then(handlers::transaction)
    |              ^^^^^^^^ implementation of `std::ops::FnOnce` is not general enough
    |
    = note: closure with signature `fn(&'0 std::boxed::Box<ruma::UserId>) -> &ruma::UserId` must implement `std::ops::FnOnce<(&std::boxed::Box<ruma::UserId>,)>`, for any lifetime `'0`...
    = note: ...but it actually implements `std::ops::FnOnce<(&std::boxed::Box<ruma::UserId>,)>`
```

which makes absolutely no sense because `handlers::transaction` is neither a closure nor does it have a `UserId` parameter.

poljar commented:
> Hope it's okay to ignore `clippy::type_complexity` in all the new places it's popped up.

Related to that, what's the reason for removing `DeviceIdBox`? Wouldn't having such type aliases alleviate the need for the type complexity ignores?

jplatte commented:
Well, I imagine lots of use cases will actually be better off using `Arc` and three type aliases (Box, Rc, Arc) per identifier type might be a bit much. Also nobody really cared about them when I asked in one of the Matrix rooms a few months back.

poljar commented:
> Well, I imagine lots of use cases will actually be better off using `Arc` and three type aliases (Box, Rc, Arc) per identifier type might be a bit much. Also nobody really cared about them when I asked in one of the Matrix rooms a few months back.

Alright, I guess we'll leave it as is.

Not sure what the warp issue is about, doesn't seem to make much sense to me either. Perhaps @johannescpk  might help out?

jplatte commented:
This has nothing to do with warp, commenting out parts of `receive_sync_response` in the base client fixes it. I'm currently narrowing it down. I suspect an issue with auto traits.

jplatte commented:
I fixed it. My [async_auto_traits](https://docs.rs/async_auto_traits) crate helped narrow it down a little bit, though actually adding those attributes to the codebase such that they can stay in seems very hard since it also resulted in different errors that I didn't fully understand.

The underlying issue seems to be with closures only being lifetime-generic in special circumstances, but not by default. I think ultimately `x` in `|x| &**x` was inferred to have the type `&'0 _` (where `0` is some concrete unnameable lifetime) instead of `for<'a> &'a _`.
In other circumstances (see last example [here](https://docs.rs/druid/latest/druid/widget/struct.Label.html#method.new)) explicitly writing `: &_` seems to fix it but this wasn't the case here. I had converted everything else to `.map(Deref::deref)` anyways though, for better readability. This was the only case I seem to have missed, and changing it too fixed the issue.

There really needs to better diagnostics for this kind of thing 👀

jplatte commented:
cc https://github.com/rust-lang/rust/issues/41078, https://github.com/rust-lang/rust/issues/86921#issuecomment-942256287

jplatte commented:
@poljar CI fixed :)

poljar reviewed (approved):
That's quite the PR, thanks.
